cv32e40s/tests/programs/custom/debug_test2/debug_test2.c

```diff
@@ -0,0 +1,3814 @@
+    .fields.rand_start_delay = 0,
+    .fields.start_delay      = 200
+  };
+
+  DEBUG_REQ_CONTROL_REG = debug_req_ctrl.raw;
+
+  while (*g_debug_status == 0) {
+    ;;
+  }
+
+  test_fail += (*g_debug_status == 1) ? 0 : 1;
+
+  if (test_fail) {
+    cvprintf(V_LOW, "\nTest: \"%s\" FAIL!\n", name);
+    return index + 1;
+  }
+  cvprintf(V_MEDIUM, "\nTest: \"%s\" OK!\n", name);
+  return 0;
+}
+
+// -----------------------------------------------------------------------------
+
+// Debug mode ecall
+void ecall_in_dmode_dbg(void) {
+  __asm__ volatile ( R"(
+    ecall
+  )");
+}
+
+// -----------------------------------------------------------------------------
+
+// Mret in dmode test entry
+uint32_t mret_in_dmode(uint32_t index, uint8_t report_name) {
+  volatile uint8_t test_fail = 0;
+  volatile debug_req_control_t debug_req_ctrl;
+
+  SET_FUNC_INFO
+
+  if (report_name) {
+    cvprintf(V_LOW, "\"%s\"", name);
+    return 0;
+  }
+
+  *g_debug_status   = 0;
+  *g_debug_test_num = 7;
+
+  debug_req_ctrl = (debug_req_control_t) {
+    .fields.value            = 1,
+    .fields.pulse_mode       = 1,
+    .fields.rand_pulse_width = 0,
+    .fields.pulse_width      = 0x1fff,
+    .fields.rand_start_delay = 0,
+    .fields.start_delay      = 200
+  };
+
+  DEBUG_REQ_CONTROL_REG = debug_req_ctrl.raw;
+
+  while (*g_debug_status == 0) {
+    ;;
+  }
+
+  test_fail += (*g_debug_status == 1) ? 0 : 1;
+
+  if (test_fail) {
+    cvprintf(V_LOW, "\nTest: \"%s\" FAIL!\n", name);
+    return index + 1;
+  }
+  cvprintf(V_MEDIUM, "\nTest: \"%s\" OK!\n", name);
+  return 0;
+}
+
+// -----------------------------------------------------------------------------
+
+// Mret in debug mode, debug portion
+// This does technically not need to be in a function
+// but for test structure and readability keeps the
+// structure used for other tests.
+void mret_in_dmode_dbg(void) {
+  __asm__ volatile ( R"(
+    mret
+  )");
+}
+
+// -----------------------------------------------------------------------------
+
+// Exception to dmode test entry
+uint32_t exception_enters_debug_mode(uint32_t index, uint8_t report_name) {
+  volatile uint8_t test_fail = 0;
+
+  SET_FUNC_INFO
+
+  if (report_name) {
+    cvprintf(V_LOW, "\"%s\"", name);
+    return 0;
+  }
+
+  *g_debug_status   = 0;
+  *g_debug_test_num = 8;
+
+  __asm__ volatile (R"(csrrs s11, dcsr, zero)");
+
+  while (*g_debug_status == 0) {
+    ;;
+  }
+
+  test_fail += (*g_debug_status == 1) ? 0 : 1;
+
+  if (test_fail) {
+    cvprintf(V_LOW, "\nTest: \"%s\" FAIL!\n", name);
+    return index + 1;
+  }
+  cvprintf(V_MEDIUM, "\nTest: \"%s\" OK!\n", name);
+  return 0;
+}
+
+// -----------------------------------------------------------------------------
+
+// Debug portion of exception -> debug entry test
+void exception_enters_debug_mode_dbg(void) {
+  volatile uint32_t dpc;
+
+  *g_debug_status = 1;
+  __asm__ volatile ( R"(
+    csrrs %[dpc], dpc, %[dpc]
+    addi %[dpc], %[dpc], 2
+    csrrw zero, dpc, %[dpc]
```

Comment:
I think the last argument in `csrrs %[dpc], dpc, %[dpc]` might be a bit wrong, using the uninitialized `volatile uint32_t dpc;` to set bits in the dpc CSR. But that CSR gets sat correctly further down so there shouldn't be any issue.

Reply:
Good catch - looks like this should have been ``` csrrs %[dpc], dpc, zero ```